lib/internal/fs/linux_recursive_watcher.js

```diff
@@ -0,0 +1,154 @@
+'use strict';
+
+const { EventEmitter } = require('events');
+const path = require('path');
+const { Symbol, ObjectKeys } = primordials;
+
+const kFSWatchStart = Symbol('kFSWatchStart');
+
+let internalSync;
+let internalPromises;
+
+function lazyLoadFsPromises() {
+  internalPromises ??= require('fs/promises');
+  return internalPromises;
+}
+
+function lazyLoadFsSync() {
+  internalSync ??= require('fs');
+  return internalSync;
+}
+
+async function traverse(dir, files = {}) {
+  const { stat, readdir } = lazyLoadFsPromises();
+
+  files[dir] = await stat(dir);
+
+  try {
+    const directoryFiles = await readdir(dir);
```

Comment:
Problem: readdir() can cause performance degradation or exhaust memory on large directory trees because it reads directory entries eagerly.

Libuv has an iterator-based API (`uv_fs_scandir()` + `uv_fs_scandir_next()`) but node doesn't expose that.

readdir() even uses the iterator API under the hood but then proceeds to consume the iterator eagerly...

Reply:
Thanks for the review! I just replaced `readdir` with `opendir`.